sound/soc/codecs/rt711-sdca.c

```diff
@@ -0,0 +1,1309 @@
+		ret = regmap_read(rt711->regmap,
+			RT711_BUF_ADDR_HID1 + offset + idx, &val);
+		if (ret < 0)
+			goto _end_btn_det_;
+		buf[idx] = val & 0xff;
+	}
+
+	if (buf[0] == 0x11) {
+		switch (buf[1] & 0xf0) {
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+		switch (buf[2]) {
+		case 0x01:
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x02:
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x04:
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x08:
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+	}
+
+_end_btn_det_:
+	/* Host is owner, so set back to device */
+	if (owner == 0)
+		/* set owener to device */
+		regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUN_HID, RT711_SDCA_ENT_HID01,
+			RT711_SDCA_CTL_HIDTX_SET_OWNER_TO_DEVICE, 0), 0x01);
+
+	return btn_type;
+}
+
+static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
+{
+	unsigned int det_mode;
+	int ret;
+
+	/* get detected_mode */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_GE49, RT711_SDCA_CTL_DETECTED_MODE, 0),
+		&det_mode);
+	if (ret < 0)
+		goto io_error;
+
+	switch (det_mode) {
+	case 0x00:
+		rt711->jack_type = 0;
+		break;
+	case 0x03:
+		rt711->jack_type = SND_JACK_HEADPHONE;
+		break;
+	case 0x05:
+		rt711->jack_type = SND_JACK_HEADSET;
+		break;
+	}
+
+	/* write selected_mode */
+	if (det_mode) {
+		ret = regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_GE49, RT711_SDCA_CTL_SELECTED_MODE, 0),
+			det_mode);
+		if (ret < 0)
+			goto io_error;
+	}
+
+	dev_dbg(&rt711->slave->dev,
+		"%s, detected_mode=0x%x\n", __func__, det_mode);
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static void rt711_sdca_jack_detect_handler(struct work_struct *work)
+{
+	struct rt711_sdca_priv *rt711 =
+		container_of(work, struct rt711_sdca_priv, jack_detect_work.work);
+	int btn_type = 0, ret;
+	unsigned int sdca_stat0, sdca_stat8;
+
+	if (!rt711->hs_jack)
+		return;
+
+	if (!rt711->component->card->instantiated)
+		return;
+
+	/* SDW_SCP_SDCA_INT_SDCA_0 is used for jack detection */
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &sdca_stat0);
+	if (ret < 0)
+		goto io_error;
+	if (sdca_stat0) {
+		ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT1, sdca_stat0);
+		if (ret < 0)
+			goto io_error;
+	}
+	if (sdca_stat0 & SDW_SCP_SDCA_INT_SDCA_0) {
+		ret = rt711_sdca_headset_detect(rt711);
+		if (ret < 0)
+			return;
+	}
+
+	/* SDW_SCP_SDCA_INT_SDCA_8 is used for button detection */
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &sdca_stat8);
+	if (ret < 0)
+		goto io_error;
+	if (sdca_stat8) {
+		ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT2, sdca_stat8);
```

Comment:
I really think you should read the SDCA_INTx registers that you need in the interrupt callback, memorise their values, clear the interrupts and use these values in the workqueue. It's confusing to clear such interrupts at a later point.

Reply:
Sure, you are right. I will move this part to the interrupt callback.